### ISSUE TITLE
Character Encoding Problem

### ISSUE THREAD

**Issue by acroedit** [remobjects/pascalscript#145]
Hello,

I've used pascal script for many years.
Recently, I want to use it in Unicode Delphi.
But it has a strange problem.

In sample3.dpr of Console folder, I've changed the script to use MessageBoxW and run the sample and it shows non-readable message.

![2017-02-13 9 22 50](https://cloud.githubusercontent.com/assets/433493/22883263/b7ad8544-f232-11e6-8e93-f2505721fd2f.png)

MessageBoxA has no problem.
Please let me know how can I fix it.

Regards,
SungDong Kim

**Comment by Valdeirsk8** [comment]
Hello @acroedit, I'm not an expert in pascal script, but I had a simular issue with Delphi Xe5 and I resolved changing the pChar to pAnsiChar, like that 
`function MessageBox(hWnd:LongInt; lbText, lpCaption:PAnsichar; uType: Longint): Longint; `
or 
`function MessageBox(hWnd:LongInt; lbText, lpCaption: AnsiString; uType: Longint): Longint;`

If you would like to know more, there are a good material here http://blog.marcocantu.com/blog/strings_unicode_delphi_2009.html

regards

**Comment by acroedit** [comment]
Thanks for the answer but It does not work with MessageBoxW@user32.dll

Regards

**Comment by pult** [comment]
function MessageBox(hWnd:LongInt; lbText, lpCaption: PChar; uType: Longint): Longint; external 'MessageBoxW@user32.dll stdcall';
OR
function MessageBox(hWnd:LongInt; lbText, lpCaption: string; uType: Longint): Longint; external 'MessageBoxW@user32.dll stdcall';

**Comment by acroedit** [comment]
Thanks @pult 

string or UnicodeString works good but PWideChar or PChar still does not work for me.

**Comment by pult** [comment]
PWideChar(and unicode PChar ) implemented in my fork.

**Comment by martijnlaan** [comment]
@acroedit I would recommend defining PS_PANSICHAR so that you yet a properly named PAnsiChar instead of PChar which like you discovered isn't actually Unicode. After that you will have AnsiString and PAnsiChar which both work together with MessageBoxA and String which works together with MessageBoxW. You will not have a PChar or PWideChar simply because it's not supported.

**Comment by acroedit** [comment]
@martijnlaan 
Thanks.

I understand that in the master ROPS, PChar is actually PAnsiChar. So this is the reason of the my problem. Is it right?

I want PWideChar to be implemented in the master ROPS and PChar to be PWideChar.

**Comment by martijnlaan** [comment]
There's already a String type which does the same thing so I doubt it's going to be added unless you do it yourself.

**Comment by carlokok** [comment]
indeed. That said, patches for full PWideChar support are always welcome.

**Comment by pult** [comment]
**@carlokok** 
https://github.com/pult/pascalscript

**Pathed:**

- - Added types: PWideChar, NativeInt, NativeUInt, RawByteString, UTF8String
- - COM: asigin interface from variant/interface (auto QueryInterface). 
    Detailed errorr on fail Interface cast. 
    IDispatchInvoke(..)
- - remove warnings/hints
- - Some fixes for CPU64 and type convertions
- - TStream fixes
- - Added exception classes (it allow filtering exception when debug app)
- - Assign TBytes/Array to string types and string types into array of IntType
- "unit-importing" added detected overide name from comment like:
    function  ObjField(FieldId: Integer): TOracleObject; overload **{_ObjField_};**
    function  ObjField(const FieldId: string): TOracleObject; overload **_{ObjFieldN};_**

**Not fixed:**
- - function pos() not work correctly for Char parameter (not string)

**Comment by carlokok** [comment]
those sound like nice fixes. You feel like doing some pull requests for them? @pult
